Added the non-constant coefficient homogenous solver

mijo2 opened the pull request:
#### References to other Issues or PRs


#### Brief description of what is fixed or changed
Added n equations linear first order non-constant coefficient homogeneous solver

#### Other comments
This solver solves the systems of ODEs of a particular form wherein the coefficient matrix of the system of ODEs has a commutative anti-derivative

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* solvers
  * Addition of a solver that handles a subclass of linear first order non-constant coefficient homogeneous systems of ODEs with any number of equations.
<!-- END RELEASE NOTES -->

mijo2 committed:
solvers: Added the non-constant homogenous solver

mijo2 commented:
@oscarbenjamin I have added the code for the solver for now. I will start to incorporate it in our code base step by step. Along with that, after the solver has been incorporated, I will find the issues that are solved by this PR and mention them later.

mijo2 committed:
solvers: Small bug fix

oscarbenjamin commented:
> @oscarbenjamin I have added the code for the solver for now. I will start to incorporate it in the code base step by step. Along with that, after the solver has been incorporated, I will find the issues that are solved by this PR and mention them later. Will that be ok?

Yes, that sounds good. Another thing to check is if there are any relevant XFAIL tests that might be fixable.

mijo2 committed:
solvers: Added the new solver into the dsolve methodology

The tests may fail as of now since they have not been modified yet,
but a basic incoporation of the new solver has been attempted in
this commit.

mijo2 commented:
I want to correct myself. Actually this solver that we are adding works only for first order system of ODEs and replaces only 4 solvers from the old code but its still better to work on this solver first since the constant coeff non-homogeneous solver replaces only one solver from the old code. 

The rest 6 solvers that can be replaced by this solver can be done so only when we add Higher order to first order reduction.

mijo2 commented:
@oscarbenjamin I think that this solver cannot replace solvers like Linear, 2 equations, Order 1, Type 5 and Linear, 3 equations, Order 1, Type 4 as, I tried to check if their coefficient matrices and its antiderivative are commutative and turns out that they aren't. If the commutative matrix isn't commutative with its anti derivative, then this solution doesn't work according to that proof mentioned in the systems roadmap. What should we do now?

oscarbenjamin commented:
> I think that this solver cannot replace solvers like Linear, 2 equations, Order 1, Type 5 and Linear, 3 equations, Order 1, Type 4

This is what I get for type 5 based on
https://docs.sympy.org/latest/modules/solvers/ode.html#linear-2-equations-order-1-type-5

```julia
In [21]: a, b = symbols('a, b')                                                                                                   

In [22]: f, g, x, y = symbols('f, g, x, y', cls=Function)                                                                         

In [23]: eq_type5 = [Eq(x(t).diff(t), f(t)*x(t) + g(t)*y(t)), 
    ...:             Eq(y(t).diff(t), a*g(t)*x(t) + (f(t) + b*g(t))*y(t))]                                                        

In [24]: eq_type5                                                                                                                 
Out[24]: 
⎡d                                 d                                            ⎤
⎢──(x(t)) = f(t)⋅x(t) + g(t)⋅y(t), ──(y(t)) = a⋅g(t)⋅x(t) + (b⋅g(t) + f(t))⋅y(t)⎥
⎣dt                                dt                                           ⎦

In [25]: from sympy.solvers.ode.systems import linear_ode_to_matrix                                                               

In [26]: ((A1, A), rhs) = linear_ode_to_matrix(eq, [x(t), y(t)], t, 1)                                                            

In [27]: A                                                                                                                        
Out[27]: 
⎡ -f(t)        -g(t)      ⎤
⎢                         ⎥
⎣-a⋅g(t)  -(b⋅g(t) + f(t))⎦

In [29]: B = A.integrate(t)                                                                                                       

In [30]: simplify(B*A - A*B)                                                                                                      
Out[30]: 
⎡0  0⎤
⎢    ⎥
⎣0  0⎦
```
So it looks to me like they do commute in that case. We probably want to avoid calling general simplify in the matching code if possible. For this example the problem is just getting the constants out of the integral signs. There are cheaper ways to do that if it's all that is needed:
```julia
In [36]: (B*A - A*B).applyfunc(expand).applyfunc(factor_terms)                                                                    
Out[36]: 
⎡0  0⎤
⎢    ⎥
⎣0  0⎦
```

oscarbenjamin commented:
I think the same applies for Linear, 3 equations, Order 1, Type 4:
https://docs.sympy.org/latest/modules/solvers/ode.html#linear-3-equations-order-1-type-4

```julia
In [48]: A                                                                                                                        
Out[48]: 
⎡-(a₁⋅f(t) + g(t))      -a₂⋅f(t)           -a₃⋅f(t)     ⎤
⎢                                                       ⎥
⎢    -b₁⋅f(t)       -(b₂⋅f(t) + g(t))      -b₃⋅f(t)     ⎥
⎢                                                       ⎥
⎣    -c₁⋅f(t)           -c₂⋅f(t)       -(c₃⋅f(t) + g(t))⎦

In [49]: B                                                                                                                        
Out[49]: 
⎡⌠                               ⌠                       ⌠             ⎤
⎢⎮ -(a₁⋅f(t) + g(t)) dt      -a₂⋅⎮ f(t) dt           -a₃⋅⎮ f(t) dt     ⎥
⎢⌡                               ⌡                       ⌡             ⎥
⎢                                                                      ⎥
⎢        ⌠               ⌠                               ⌠             ⎥
⎢    -b₁⋅⎮ f(t) dt       ⎮ -(b₂⋅f(t) + g(t)) dt      -b₃⋅⎮ f(t) dt     ⎥
⎢        ⌡               ⌡                               ⌡             ⎥
⎢                                                                      ⎥
⎢        ⌠                       ⌠               ⌠                     ⎥
⎢    -c₁⋅⎮ f(t) dt           -c₂⋅⎮ f(t) dt       ⎮ -(c₃⋅f(t) + g(t)) dt⎥
⎣        ⌡                       ⌡               ⌡                     ⎦

In [50]: simplify(B*A - A*B)                                                                                                      
Out[50]: 
⎡0  0  0⎤
⎢       ⎥
⎢0  0  0⎥
⎢       ⎥
⎣0  0  0⎦

In [51]: (B*A - A*B).applyfunc(expand).applyfunc(factor_terms)                                                                    
Out[51]: 
⎡0  0  0⎤
⎢       ⎥
⎢0  0  0⎥
⎢       ⎥
⎣0  0  0⎦
```

mijo2 commented:
@oscarbenjamin I didn't simplify the expressions when I was verifying. The expressions looked unequal at the surface. I would add the code that you mentioned.

mijo2 commented:
@oscarbenjamin Some of the old test cases have extremely complicated solutions using our method partially because we are using ```integrate``` to get the antiderivative of the coefficient matrix.

oscarbenjamin commented:
Can you give an example?

mijo2 commented:
> Can you give an example?
```
    x, y, z = symbols('x, y, z', cls=Function)
    t = Symbol('t')

    f = t**3 + log(t)
    g = t**2 + sin(t)
    eq1 = (Eq(diff(x(t),t),(4*f+g)*x(t)-f*y(t)-2*f*z(t)), Eq(diff(y(t),t),2*f*x(t)+(f+g)*y(t)-2*f*z(t)), Eq(diff(z(t),t),5*f*x(t)+f*y(t)+(-3*f+g)*z(t)))
```

The above equation ```eq1``` has a coefficient matrix which has an antiderivative whose exponential is very complicated.

oscarbenjamin commented:
So here is the matrix:
```julia
In [29]:     x, y, z = symbols('x, y, z', cls=Function) 
    ...:     t = Symbol('t') 
    ...:  
    ...:     f = t**3 + log(t) 
    ...:     g = t**2 + sin(t) 
    ...:     eq1 = (Eq(diff(x(t),t),(4*f+g)*x(t)-f*y(t)-2*f*z(t)), Eq(diff(y(t),t),2*f*x(t)+(f+g)*y(t)-2*f*z(t)), Eq(diff(z(t),t),5*f*x(t)+f*y(
    ...: t)+(-3*f+g)*z(t)))                                                                                                                    

In [30]: from sympy.solvers.ode.systems import linear_ode_to_matrix                                                                            

In [31]: (A1, A2), rhs = linear_ode_to_matrix(eq1, [x(t), y(t), z(t)], t, 1)                                                                   

In [32]: A2                                                                                                                                    
Out[32]: 
⎡ ⎛   3    2                    ⎞           3                                3                     ⎤
⎢-⎝4⋅t  + t  + 4⋅log(t) + sin(t)⎠          t  + log(t)                    2⋅t  + 2⋅log(t)          ⎥
⎢                                                                                                  ⎥
⎢        ⎛   3           ⎞          ⎛ 3    2                  ⎞              3                     ⎥
⎢       -⎝2⋅t  + 2⋅log(t)⎠         -⎝t  + t  + log(t) + sin(t)⎠           2⋅t  + 2⋅log(t)          ⎥
⎢                                                                                                  ⎥
⎢        ⎛   3           ⎞                 ⎛ 3         ⎞          ⎛     3    2                    ⎞⎥
⎣       -⎝5⋅t  + 5⋅log(t)⎠                -⎝t  + log(t)⎠         -⎝- 3⋅t  + t  - 3⋅log(t) + sin(t)⎠⎦
```
Firstly note that the factored out minus sign slows down integration quite a lot. It's probably worth changing `linear_ode_to_matrix` to distribute that minus:
```julia
In [33]: %time ok=expand(A2).integrate(t)                                                                                                      
CPU times: user 430 ms, sys: 2.79 ms, total: 432 ms
Wall time: 432 ms

In [34]: %time ok=A2.integrate(t)                                                                                                              
CPU times: user 20.5 s, sys: 98.5 ms, total: 20.6 s
Wall time: 21 s
```
Secondly yes, the exponential of the integral of the matrix is complicated. I did expect that we would reach this point where actually getting the matrix exponential to compute something reasonable would get complicated.

Note that we can use `exp(A + B) = exp(A)*exp(B)` if `A` and `B` commute so e.g. in this case I wonder if we can get a nicer solution by splitting the matrix into a matrix of polys in `t` and a matrix of `log` terms and a matrix of `sin` terms. The `sin` matrix would be diagonal and I expect that the exp of the matrix of log terms would simplify etc.

oscarbenjamin commented:
I think there should be a way to relate the transformation here:
https://docs.sympy.org/latest/modules/solvers/ode.html#linear-3-equations-order-1-type-4
in terms of matrix exponentials.

oscarbenjamin commented:
It looks like the `eq1` example above is solved by linear 3eq order1 type4 which gives a solution for it. However checksysodesol doesn't think the solution is correct and looking at the ODE it doesn't seem to match the pattern.

mijo2 commented:
> It's probably worth changing linear_ode_to_matrix to distribute that minus

Will it be wise to use something like ```Ai.applyfunc(expand)``` in ```linear_ode_to_matrix```?

oscarbenjamin commented:
Normally a minus sign distributes automatically. It might be worth checking why that hasn't happened:
```
In [1]: -(x+y)                                                                                                                                 
Out[1]: -x - y
```

mijo2 commented:
> It might be worth checking why that hasn't happened

Ok I will go through a debugger and try to figure that out why that happens.

mijo2 committed:
solvers: Moved the test cases which are solved by the new solver

The test cases which are solved by the new solver are moved from
test_ode.py to test_systems.py

mijo2 committed:
solvers: Removed unused imports and variables

mijo2 commented:
> It might be worth checking why that hasn't happened

So I did a little checking around and I found this:

```
In [51]: Eq(x(t).diff(t), -f*x(t))                                                                                                                                                                          
Out[51]: Eq(Derivative(x(t), t), (-t**3 - log(t))*x(t))

In [52]: Eq(diff(x(t), t), -f*x(t))                                                                                                                                                                         
Out[52]: Eq(Derivative(x(t), t), (-t**3 - log(t))*x(t))

In [53]: Eq(diff(x(t), t), -f*x(t) -2*g*z(t))                                                                                                                                                               
Out[53]: Eq(Derivative(x(t), t), -(2*t**2 + 2*sin(t))*z(t) + (-t**3 - log(t))*x(t))

In [54]: Eq(diff(x(t), t), -f*x(t) - 2*g*z(t))                                                                                                                                                              
Out[54]: Eq(Derivative(x(t), t), -(2*t**2 + 2*sin(t))*z(t) + (-t**3 - log(t))*x(t))

In [55]: Eq(diff(x(t), t), -f*x(t) - 2*(t**2 + log(t))*z(t))                                                                                                                                                
Out[55]: Eq(Derivative(x(t), t), -(2*t**2 + 2*log(t))*z(t) + (-t**3 - log(t))*x(t))

In [56]: -f*x(t) - 2*(t**2 + log(t))*z(t)                                                                                                                                                                   
Out[56]: -(2*t**2 + 2*log(t))*z(t) + (-t**3 - log(t))*x(t)
```

Here, you can see that when we do something like: ```-f*x(t) - 2*(t**2 + log(t))*z(t)```, it expands the first term ```f``` but doesn't expand the second term.

mijo2 committed:
solvers: Updated a doctest

mijo2 commented:
> I wonder if we can get a nicer solution by splitting the matrix into a matrix of polys in t and a matrix of log terms and a matrix of sin terms. The sin matrix would be diagonal and I expect that the exp of the matrix of log terms would simplify etc.

Do we have to create a new function for this from scratch or are there any functions to do the same?

oscarbenjamin commented:
> Do we have to create a new function for this from scratch or are there any functions to do the same?

I don't think there are any functions for this although presumably the type4 solver has some code for detecting the form.

oscarbenjamin commented:
> Here, you can see that when we do something like: `-f*x(t) - 2*(t**2 + log(t))*z(t)`, it expands the first term `f` but doesn't expand the second term.

The 2-arg Mul distributes a rational which breaks associativity e.g.:
```julia
In [2]: a = Integer(2)                                                                                                                         

In [3]: b = x + y                                                                                                                              

In [4]: c = z                                                                                                                                  

In [5]: a*(b*c)                                                                                                                                
Out[5]: 2⋅z⋅(x + y)

In [6]: (a*b)*c                                                                                                                                
Out[6]: z⋅(2⋅x + 2⋅y)
```

mijo2 commented:
> The 2-arg Mul distributes a rational which breaks associativity e.g.:
So, what should we do in this situation to fix it?

mijo2 commented:
> in this case I wonder if we can get a nicer solution by splitting the matrix into a matrix of polys in t and a matrix of log terms and a matrix of sin terms. The sin matrix would be diagonal and I expect that the exp of the matrix of log terms would simplify etc.

So, I was trying this through the same example and made a few matrices that when summed up gave our original matrix. The matrix which was made up of the term ```t*log(t)``` took the most time to compute, almost as much as the original coefficient matrix and I think this may be one of the reasons why the computation is slow.

mijo2 committed:
solvers: Added markers for future tasks in this PR.

mijo2 committed:
solvers: Updated the constant coeff homogeneous solver docstring

The docstring of the old solver which was recently changed was
updated.

oscarbenjamin commented:
> So, what should we do in this situation to fix it?

We need to find the cause of it. I think it's something to do with the way that `linear_ode_to_matrix` subtracts from the two sides of the equation.

oscarbenjamin commented:
Actually it looks like this is just what happens with `linear_eq_to_matrix`:
```
In [3]: linear_eq_to_matrix([-((t+y)*x)], [x])                                                                                                 
Out[3]: ([-(t + y)], [0])
```
I wouldn't use `expand`. If integrate is slow because of this then maybe an issue can be opened with integrals.

oscarbenjamin commented:
> I wouldn't use `expand`.

Although if `expand` helps with computing the matrix exponential then that is different.

mijo2 commented:
> If integrate is slow because of this then maybe an issue can be opened with integrals.
> Although if `expand` helps with computing the matrix exponential then that is different.

So, the usage of `expand` helps for both, integrating as well as computing the exponential. Here is a simple example demonstrating the same:

```
In [1]: from sympy import * 
   ...: from sympy.abc import * 
   ...: from sympy.calculus import * 
   ...: from sympy.calculus.util import * 
   ...: from sympy.solvers.ode.tests.test_ode import * 
   ...: from sympy.solvers.ode.ode import * 
   ...: from sympy.solvers.ode.systems import *                                                                                                                                                             

In [2]: eqs = [Eq(f(x).diff(x), x*f(x) - (x + x**2)*g(x)), Eq(g(x).diff(x), (x + x**2)*f(x) + x*g(x) )]                                                                                                     

In [3]: (A1, A), rhs = linear_ode_to_matrix(eqs, [f(x), g(x)], x, 1)                                                                                                                                        

In [4]: A                                                                                                                                                                                                   
Out[4]: 
Matrix([
[         -x, x**2 + x],
[-(x**2 + x),       -x]])

In [5]: A_exp = A.applyfunc(expand)                                                                                                                                                                         

In [6]: from sympy.solvers.ode.systems import _is_commutative_anti_derivative 

In [8]: %time B, is_commuting = _is_commutative_anti_derivative(A, x)                                                                                                                                       
CPU times: user 24.9 ms, sys: 24 µs, total: 25 ms
Wall time: 24.7 ms

In [9]: %time B_expand, is_commuting = _is_commutative_anti_derivative(A_exp, x)                                                                                                                            
CPU times: user 10.8 ms, sys: 112 µs, total: 10.9 ms
Wall time: 12.4 ms

In [11]: is_commuting                                                                                                                                                                                       
Out[11]: True

In [12]: %time B_exp = B.exp()                                                                                                                                                                              
CPU times: user 547 ms, sys: 3.78 ms, total: 551 ms
Wall time: 550 ms

In [13]: %time B_expand_exp = B_expand.exp()                                                                                                                                                                
CPU times: user 181 ms, sys: 3.62 ms, total: 185 ms
Wall time: 184 ms

```

mijo2 committed:
solvers: Added new docstring for the solver introduced

mijo2 committed:
doc: Added the new solver in the documentation

The new solver was added in the documentation as Linear n equations,
Order 1, Type 3 in the systems section of ode.rst. The build was also
checked which resulted in success.

mijo2 committed:
solvers: Added new checking mechanism for handling general case

New checking code was added in sysode_linear_neq_order1 to handle
the two solvers that are added in systems.py

mijo2 committed:
solvers: Removed _linear_neq_order1_type1 solver

Removed _linear_neq_order1_type1 solver from sysode_linear_3eq_order1

mijo2 committed:
solvers: Updated the doctest for classify_sysode

mijo2 commented:
@oscarbenjamin I would start contributing for this PR soon. As of now, I am setting up my blog on GitHub pages and writing my first blog post.

mijo2 committed:
solvers: Moved a test case from test_ode.py to test_systems.py

Moved the test case which was now being handled by the new solver.
This test case is still slow and thus the correct version of the
test case given the new solver was used.

mijo2 commented on sympy/solvers/ode/tests/test_ode.py:
```diff
@@ -160,18 +160,8 @@ def test_linear_2eq_order1_type2_fixme_check():
-
+# This test is not fixed by the new solver because the
+# coefficient matrix's antiderivative doesn't commute.
 def test_linear_2eq_order1_type4_broken():
```
@oscarbenjamin I think this is a faulty test case. This test case doesn't have a commutative antiderivative and doesn't even have the form described in Linear, 2 equations, Order 1, Type 4. Should I remove this test case altogether?

oscarbenjamin replied:
There can be solutions to systems that don't have a commutative antiderivative so that is not a reason for removing a test case if it is currently solved by one of the solvers.

However in this case it looks as if this example should not match any of the existing solvers and the solution given by type4 is incorrect.

Yes I think it's fine to remove this.

If we knew the correct answer for the system of ODEs then it would be good to add that as an XFAIL test instead though.

oscarbenjamin replied:
I can't find what the solution should be in this case. It's possible that there is no solution in ordinary functions. I think it's best to remove this example.

mijo2 replied:
> There can be solutions to systems that don't have a commutative antiderivative so that is not a reason for removing a test case if it is currently solved by one of the solvers.

Ok, I will keep this in mind from now on when dealing with XFail tests.

>  I think it's best to remove this example.

Sure thing, I will remove this example.

mijo2 committed:
solvers, doc: Removed the unwanted solvers

4 of the solvers that were to be removed from the source code were
removed along with their documentation.

mijo2 committed:
solvers: Identified the reason why some expressions wont collect

Some of the solutions don't have correctly collected exp terms
and that can be avoided using expand. For now, only some markers
are added identifying the problem. Also, a new test case is added.

mijo2 committed:
solvers: Added some more test cases in test_systems.py

Some test cases were added for the new solver where the
coefficient matrix has all the elements same.

mijo2 commented on sympy/solvers/ode/systems.py:
```diff
@@ -486,6 +446,79 @@ def _canonical_equations(eqs, funcs, t):
+    # the powers where it isn't possible.
+    # For example: collect(a*(x+2), x) == a*(x + 2) but
+    # collect(expand(a*(x+2)), x) == a*x + 2*a
+    sol_vector = [collect(s, ordered(s.atoms(exp)), exact=True) for s in sol_vector]
```
@oscarbenjamin As mentioned in this marker, some of the expressions are not being collected properly. Can we use ```expand``` function here as a workaround?

oscarbenjamin replied:
Can you give an example of what `sol_vector` is here?

mijo2 replied:
For ```dsolve(eqs)``` where ```eqs = [Eq(f(x).diff(x), x * f(x) + g(x)), Eq(g(x).diff(x), -f(x) + x * g(x))]```, 

```
sol_vector before collect is Matrix([[C1*(exp(x**2/2 - I*x)/2 + exp(x**2/2 + I*x)/2)
 + C2*(I*exp(x**2/2 - I*x)/2 - I*exp(x**2/2 + I*x)/2)], 
[C1*(-I*exp(x**2/2 - I*x)/2 + I*exp(x**2/2 + I*x)/2) 
+ C2*(exp(x**2/2 - I*x)/2 + exp(x**2/2 + I*x)/2)]])

sol_vector after collect is [C1*(exp(x**2/2 - I*x)/2 + exp(x**2/2 + I*x)/2)
+ C2*(I*exp(x**2/2 - I*x)/2 - I*exp(x**2/2 + I*x)/2),
C1*(-I*exp(x**2/2 - I*x)/2 + I*exp(x**2/2 + I*x)/2) + C2*(exp(x**2/2 - I*x)/2 + exp(x**2/2 + I*x)/2)]
```

oscarbenjamin replied:
We could use `expand_mul` I guess. In this example ideally the solution would come out like this:
```julia
In [31]: M=ImmutableMatrix([[C1*(exp(x**2/2 - I*x)/2 + exp(x**2/2 + I*x)/2) 
    ...:  + C2*(I*exp(x**2/2 - I*x)/2 - I*exp(x**2/2 + I*x)/2)],  
    ...: [C1*(-I*exp(x**2/2 - I*x)/2 + I*exp(x**2/2 + I*x)/2)  
    ...: + C2*(exp(x**2/2 - I*x)/2 + exp(x**2/2 + I*x)/2)]])                                                                                   

In [32]: (M / exp(x**2/2)).expand().rewrite(cos).expand() * exp(x**2/2)                                                                        
Out[32]: 
⎡                          2 ⎤
⎢                         x  ⎥
⎢                         ── ⎥
⎢                         2  ⎥
⎢(C₁⋅cos(x) + C₂⋅sin(x))⋅ℯ   ⎥
⎢                            ⎥
⎢                           2⎥
⎢                          x ⎥
⎢                          ──⎥
⎢                          2 ⎥
⎣(-C₁⋅sin(x) + C₂⋅cos(x))⋅ℯ  ⎦
```

oscarbenjamin replied:
Another possibility is that maybe we should make `x` real:
```julia
In [115]: x = Symbol('x', real=True)                                                                                                           

In [116]: B = Matrix([[-x**2/2, -x], [x, -x**2/2]])                                                                                            

In [117]: B.exp()                                                                                                                              
Out[117]: 
⎡   2              2        ⎤
⎢ -x             -x         ⎥
⎢ ────           ────       ⎥
⎢  2              2         ⎥
⎢ℯ    ⋅cos(x)  -ℯ    ⋅sin(x)⎥
⎢                           ⎥
⎢   2             2         ⎥
⎢ -x            -x          ⎥
⎢ ────          ────        ⎥
⎢  2             2          ⎥
⎣ℯ    ⋅sin(x)  ℯ    ⋅cos(x) ⎦
```

mijo2 replied:
> We could use expand_mul I guess. In this example ideally the solution would come out like this:

@oscarbenjamin Yes, ```expand_mul``` would be better since that's what we require for collecting expressions that are not being collected due to the brackets.

mijo2 replied:
> Another possibility is that maybe we should make x real

I would add a test case regarding this.

mijo2 committed:
solvers: Added expand_mul in the new solver

The new solver now expands some terms in order to collect the
exponential terms.

mijo2 committed:
solvers: Added test case for when x is real

A test case was modified and added when the independent variable
was real.

mijo2 committed:
solvers: Added regression test case for issue #18562

This test case demonstrates the case when y in the issue's example
is independent of x. When it is dependent on x, then for most of the
cases, the coefficient matrix and its anti derivative don't commute.

mijo2 committed:
Merge branch 'master' into non_const_homogeneous

mijo2 commented:
@oscarbenjamin I wasn't able to find any issues that are fixed by the addition of this PR but there are many issues for non-homogeneous cases of systems of ODEs that will be fixed once we add those solvers.

mijo2 commented:
@oscarbenjamin There is still one test case that has a very complicated with the new solver but overall, this solver effectively replaces four of the old solvers along with adding more coverage of the systems of ODEs that can be solved by the library. But, one thing remains:

Should we name the new solvers as ```_linear_neq_order1_type{x}``` format where we currently have 4 solvers in line that will be added in total. PR #18720 added and updated ```_linear_neq_order1_type1``` and this PR is adding ```_linear_neq_order1_type3``` where type 2 and type 4 are the non-homogeneous equivalents. This will make the names of the solvers less descriptive.

mijo2 committed:
solvers: Removed lots of code related to old solvers

Solvers that were removed had all the matching code for them still
in the module and those were removed in this commit. The whole
sysode_linear_3eq_order1 function was removed since it was no longer
needed.

oscarbenjamin commented:
> There is still one test case that has a very complicated with the new solver

Which case is that?

I'm not sure what the solvers should be named. I imagine a big rearrangement once the old solvers and matching code are gone.

mijo2 commented:
> Which case is that?

https://github.com/sympy/sympy/blob/89df39fa2daf7c9b9e3877542c8e3d0d7d0855b3/sympy/solvers/ode/tests/test_systems.py#L749-L759

This is the test case. 

> I'm not sure what the solvers should be named.

I think it would be wise to name them using types, making the function names smaller.

mijo2 committed:
doc, solvers: Updated the names of the general case solvers

The general solvers that are added have their names updated for now,
with names containing type 1 and type 3 to reduce the length of the
names of the solvers.

mijo2 committed:
solvers: Removed some of the markers from the tasks to be done

Some markers for which the associated tasks were completed, were
removed in this commit.

mijo2 commented:
@oscarbenjamin Regarding this PR, I think only the last test case in ```test_systems.py``` is remaining to be tackled(mentioned in the previous [comment](https://github.com/sympy/sympy/pull/19185#issuecomment-627270992). Rest of the work of replacing the old solvers and adding the new solver is mostly taken care of

oscarbenjamin commented:
I don't think that there is any immediate problem with that case. The solution is complicated but maybe it needs to be. Note that the previous solution in the tests is not correct.

There are possible ways to simplify this sort of matrix exponential but those can be improvements for the future.

mijo2 commented:
@oscarbenjamin I agree that we can work on simplifying complicated solutions later. But, this test case is failing in the build even though it passes when I try it locally(due to the reason that the test case takes too much time). So, for now, can I add it as a ```XFail``` test case?

oscarbenjamin commented:
Is it `dsolve` or `checkodesol` that takes too long?

If it's `checkodesol` then split that out into a separate test that is marked as `slow` and uses `skip("too slow for Travis")`. It shouldn't be marked as xfail if it actually passes.

mijo2 commented:
> Is it `dsolve` or `checkodesol` that takes too long?

Its `checkodesol`, `dsolve` now solves the problem in roughly 13s.

> If it's `checkodesol` then split that out into a separate test that is marked as `slow` and uses `skip("too slow for Travis")`. It shouldn't be marked as xfail if it actually passes.

Sure, will do that.

mijo2 committed:
solvers: Marked a test case as too slow

Added a clause in a test case which is too slow for travis but it
passes in a local machine.

mijo2 commented:
@oscarbenjamin All the tests have passed. I think it is ready to be merged.

oscarbenjamin commented:
I don't know why Travis shows the tests passing but github says the checks aren't completed. I have a couple of suggestions so perhaps it will be fixed after the next push...

oscarbenjamin commented on sympy/solvers/ode/systems.py:
```diff
@@ -169,7 +171,10 @@ def linear_ode_to_matrix(eqs, funcs, t, order):
         except NonlinearError:
             raise ODENonlinearError("The system of ODEs is nonlinear.")
 
+        Ai = Ai.applyfunc(expand)
```
Maybe this should just be `expand_mul`. What is this needed for?

mijo2 replied:
Ok, I will update this, it's not needed anymore actually.

oscarbenjamin commented on sympy/solvers/ode/tests/test_systems.py:
```diff
@@ -604,3 +674,99 @@ def test_sysode_linear_neq_order1():
+    #         Eq(z(t), (C1*exp(-2*Integral(t**3 + log(t), t)) + C2*cos(sqrt(3)*Integral(t**3 + log(t), t)) - \
+    #         C3*sin(sqrt(3)*Integral(t**3 + log(t), t)))*exp(Integral(-t**2 - sin(t), t)))]
+
+    dsolve_sol = dsolve(eq1)
```
If `dsolve` here only takes 13 seconds then I think this test function should be split into two:

1. A test for `dsolve` marked as `@slow`.
2. A test for checkodesol that is marked as `@slow` but also "too slow for Travis".

mijo2 replied:
@oscarbenjamin Sorry, I misunderstood what you suggested earlier. I would update it as you are suggesting now.

mijo2 committed:
solvers: Splitted a test case into two parts

One test case for too slow for travis when used checksysodesol
so the dsolve part was done separately without the skipping and
the checksysodesol part is being done separately with skipping.

mijo2 committed:
solvers: Updated the applyfun method for matrix

In linear_ode_to_matrix, for every matrix found, the expand_mul
method is applied instead of expand since its faster to do so.

oscarbenjamin commented:
This is excellent. Well done for getting it done so quickly!

mijo2 commented:
It wouldn't have been possible without your feedback and reviews!
